### ISSUE TITLE
Yarn errors while building jupyter lab

### ISSUE THREAD

**Issue by jhossbach** [jupyterlab/jupyterlab#16393]
## Description

Came across this after I couldn't update some extensions that were recommended to build.
I completely uninstalled and reinstalled jupyter and all extensions (running `jupyter lab clean --all`). Running `jupyter lab build` afterwards gives this output found below:

<details><summary>Command Line Output</summary>
<pre>
[LabBuildApp] JupyterLab 4.2.1
[LabBuildApp] Building in /home/juli/.local/share/jupyter/lab
[LabBuildApp] `sys_prefix` level settings are read-only, using `user` level for migration to `lockedExtensions`
[LabBuildApp] Building jupyterlab assets (production, minimized)
[LabBuildApp] ERROR | Build failed.
Troubleshooting: If the build failed due to an out-of-memory error, you
may be able to fix it by disabling the `dev_build` and/or `minimize` options.

If you are building via the `jupyter lab build` command, you can disable
these options like so:

jupyter lab build --dev-build=False --minimize=False

You can also disable these options for all JupyterLab builds by adding these
lines to a Jupyter config file named `jupyter_config.py`:

c.LabBuildApp.minimize = False
c.LabBuildApp.dev_build = False

If you don't already have a `jupyter_config.py` file, you can create one by
adding a blank file of that name to any of the Jupyter config directories.
The config directories can be listed by running:

jupyter --paths

Explanation:

- `dev-build`: This option controls whether a `dev` or a more streamlined
`production` build is used. This option will default to `False` (i.e., the
`production` build) for most users. However, if you have any labextensions
installed from local files, this option will instead default to `True`.
Explicitly setting `dev-build` to `False` will ensure that the `production`
build is used in all circumstances.

- `minimize`: This option controls whether your JS bundle is minified
during the Webpack build, which helps to improve JupyterLab's overall
performance. However, the minifier plugin used by Webpack is very memory
intensive, so turning it off may help the build finish successfully in
low-memory environments.

/usr/lib/python3.12/site-packages/jupyterlab/debuglog.py:54: UserWarning: An error occurred.
  warnings.warn("An error occurred.")
/usr/lib/python3.12/site-packages/jupyterlab/debuglog.py:55: UserWarning: RuntimeError: npm dependencies failed to install
  warnings.warn(msg[-1].strip())
/usr/lib/python3.12/site-packages/jupyterlab/debuglog.py:56: UserWarning: See the log file for details: /tmp/jupyterlab-debug-nsl_dzkv.log
  warnings.warn(f"See the log file for details: {log_path!s}")
</pre>
</details>

This is the mentioned debug `/tmp/jupyterlab-debug-nsl_dzkv.log` (omitting the `can't be found in the cache and  will be fetched from the remote repository` lines from yarn:

<details><summary>Log file</summary>
<pre>
[LabBuildApp] Building in /home/juli/.local/share/jupyter/lab
[LabBuildApp] Node v22.2.0

[LabBuildApp] Yarn configuration loaded.
[LabBuildApp] `sys_prefix` level settings are read-only, using `user` level for migration to `lockedExtensions`
[LabBuildApp] Building jupyterlab assets (production, minimized)
[LabBuildApp] > node /usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js install
[LabBuildApp] ➤ YN0000: ┌ Resolution step
➤ YN0002: │ @jupyterlab/cell-toolbar-extension@npm:4.2.1 doesn't provide react (pc6645), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/completer@npm:4.2.1 doesn't provide react (p2dec1), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/console-extension@npm:4.2.1 doesn't provide react (p2c3d5), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/csvviewer-extension@npm:4.2.1 doesn't provide react (p12f72), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/debugger-extension@npm:4.2.1 doesn't provide react (p1c718), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/documentsearch-extension@npm:4.2.1 doesn't provide react (pbc3ea), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/extensionmanager-extension@npm:4.2.1 doesn't provide react (p8ab60), requested by @jupyterlab/settingregistry
➤ YN0060: │ @jupyterlab/extensionmanager@npm:4.2.1 provides react (pba009) with version 18.2.0, which doesn't satisfy what react-paginate requests
➤ YN0002: │ @jupyterlab/filebrowser-extension@npm:4.2.1 doesn't provide react (pa0283), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/fileeditor-extension@npm:4.2.1 doesn't provide react (p8d9c9), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/htmlviewer-extension@npm:4.2.1 doesn't provide react (p60f0e), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/json-extension@npm:4.2.1 doesn't provide @types/react (p29fd3), requested by react-json-tree
➤ YN0002: │ @jupyterlab/mainmenu-extension@npm:4.2.1 doesn't provide react (p0fc58), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/markdownviewer-extension@npm:4.2.1 doesn't provide react (p88da2), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/metadataform-extension@npm:4.2.1 doesn't provide react (p762cb), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/metadataform@npm:4.2.1 doesn't provide @rjsf/utils (paa218), requested by @rjsf/core
➤ YN0002: │ @jupyterlab/metadataform@npm:4.2.1 doesn't provide @rjsf/utils (ped511), requested by @rjsf/validator-ajv8
➤ YN0002: │ @jupyterlab/services@npm:7.2.1 doesn't provide react (p3c2ea), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/settingeditor-extension@npm:4.2.1 doesn't provide react (p286b3), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/statusbar-extension@npm:4.2.1 doesn't provide react (pd6fae), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/terminal-extension@npm:4.2.1 doesn't provide react (p1fd3f), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/toc-extension@npm:6.2.1 doesn't provide react (pb363b), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/translation-extension@npm:4.2.1 doesn't provide react (p44b99), requested by @jupyterlab/settingregistry
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
➤ YN0000: └ Completed in 0s 212ms
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed in 0s 557ms
➤ YN0000: ┌ Link step
➤ YN0007: │ @fortawesome/fontawesome-free@npm:5.15.4 must be built because it never has been before or the last one failed
➤ YN0007: │ core-js@npm:3.30.2 must be built because it never has been before or the last one failed
➤ YN0009: │ @fortawesome/fontawesome-free@npm:5.15.4 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-0e5efe2d/build.log)
➤ YN0009: │ core-js@npm:3.30.2 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-9e2dacaf/build.log)
➤ YN0000: └ Completed in 8s 855ms
➤ YN0000: Failed with errors in 9s 688ms

[LabBuildApp] npm dependencies failed to install
[LabBuildApp] ERROR | Build failed.
Troubleshooting: If the build failed due to an out-of-memory error, you
may be able to fix it by disabling the `dev_build` and/or `minimize` options.

If you are building via the `jupyter lab build` command, you can disable
these options like so:

jupyter lab build --dev-build=False --minimize=False

You can also disable these options for all JupyterLab builds by adding these
lines to a Jupyter config file named `jupyter_config.py`:

c.LabBuildApp.minimize = False
c.LabBuildApp.dev_build = False

If you don't already have a `jupyter_config.py` file, you can create one by
adding a blank file of that name to any of the Jupyter config directories.
The config directories can be listed by running:

jupyter --paths

Explanation:

- `dev-build`: This option controls whether a `dev` or a more streamlined
`production` build is used. This option will default to `False` (i.e., the
`production` build) for most users. However, if you have any labextensions
installed from local files, this option will instead default to `True`.
Explicitly setting `dev-build` to `False` will ensure that the `production`
build is used in all circumstances.

- `minimize`: This option controls whether your JS bundle is minified
during the Webpack build, which helps to improve JupyterLab's overall
performance. However, the minifier plugin used by Webpack is very memory
intensive, so turning it off may help the build finish successfully in
low-memory environments.

[LabBuildApp] Traceback (most recent call last):

[LabBuildApp]   File "/usr/lib/python3.12/site-packages/jupyterlab/debuglog.py", line 45, in debug_logging
    yield

[LabBuildApp]   File "/usr/lib/python3.12/site-packages/jupyterlab/labapp.py", line 208, in start
    raise e

[LabBuildApp]   File "/usr/lib/python3.12/site-packages/jupyterlab/labapp.py", line 199, in start
    build(

[LabBuildApp]   File "/usr/lib/python3.12/site-packages/jupyterlab/commands.py", line 512, in build
    return handler.build(
           ^^^^^^^^^^^^^^

[LabBuildApp]   File "/usr/lib/python3.12/site-packages/jupyterlab/commands.py", line 745, in build
    raise RuntimeError(msg)

[LabBuildApp] RuntimeError: npm dependencies failed to install

[LabBuildApp] Exiting application: JupyterLab

</pre>
</details>

The logs seem to indicate two errors, one in `@fortawesome/fontawesome-free@npm:5.15.4` and `core-js@npm:3.30.2`, see the logs below:

<details><summary>core-js@npm log</summary>
<pre>
# This file contains the result of Yarn building a package (core-js@npm:3.30.2)
# Script name: postinstall

Error: spawn ETXTBSY
    at ChildProcess.spawn (node:internal/child_process:421:11)
    at Object.spawn (node:child_process:760:9)
    at oU (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:4:7031)
    at qf.implementation (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:392:17799)
    at qf.exec (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:395:1585)
    at qf.run (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:395:1756)
    at pV (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:401:6331)
    at async Ake (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:403:16)
    at async o (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:403:146)
    at async lke (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:404:14)
</pre>
</details>

<details><summary>@fortawesome/fontawesome-free@npm log</summary>
<pre>
# This file contains the result of Yarn building a package (@fortawesome/fontawesome-free@npm:5.15.4)
# Script name: postinstall

Error: spawn ETXTBSY
    at ChildProcess.spawn (node:internal/child_process:421:11)
    at Object.spawn (node:child_process:760:9)
    at oU (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:4:7031)
    at qf.implementation (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:392:17799)
    at qf.exec (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:395:1585)
    at qf.run (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:395:1756)
    at pV (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:401:6331)
    at async Ake (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:403:16)
    at async o (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:403:146)
    at async lke (/usr/lib/python3.12/site-packages/jupyterlab/staging/yarn.js:404:14)
</pre>
</details>

Any help would be appreciated!

## Context

- Operating System and version: `Arch linux with kernel 6.9.2-arch1-1`
- ~Browser and version:~
- JupyterLab version: `4.2.1`
- NPM version: `10.8.0`
- Node version: `v22.2.0`

**Comment by krassowski** [comment]
In principle the best approach is to never install extensions via `jupyter labextension install` - this command is deprecated and will be removed/renamed in the future (nowadays it is intended for administrators of deployments who want to bundle/build the extensions into the codebase, not for users). Normal users should install pre-built extensions whenever possible (e.g. with `pip install`). If an extension does not offer a prebuilt installation method it is quite likely that it was not upgraded to JupyterLab 4.x and may not work in it either way.

If you want to continue, I would suggest trying with Node 20.x rather than 22.x (we are testing against 20.x and it often takes time for things to get fixed after a new release of Node).

What specific extensions are you trying to install?

**Comment by jhossbach** [comment]
I am starting from a clean state without any extensions and not trying to, after running `jupyter lab clean --all` and starting jupyter lab I get this on the browser
![image](https://github.com/jupyterlab/jupyterlab/assets/50084485/c4bff638-79e9-48e2-9806-6361022b413a)

Running `jupyter lab build` is when the mentioned error happens.
As for managing extensions, I installed jupyterlab and, previously, all extensions using the package manager of arch (`pacman`).

I also checked `/usr/share/jupyter/labextensions` for any extensions that I missed but there is only `jupyterlab_pygments` which gets installed with pacmans `jupyterlab` package.

**Comment by jhossbach** [comment]
Any ideas on how to debug this further? Is it possible to have more verboes yarn logs?

**Comment by jhossbach** [comment]
Okay, seems like overall I have (had) two problems: The first one was recommended builds failing (not sure about the specific error). In order to fix this I wanted to start without any extensions and then reintroduce each extension to find a possible culprit, but by mistakenly running `jupyter lab clean --all` I also removed the application assets.

So I tried to install node v20 using `pacman`, and tl;dr do not try to downgrade via `pacman`, forcing all dependencies to fit will break `pacman` itself (I shouldn't have done this on a friday...)

Using `nvm` from the AUR allowed me to install node v20, after which I was successfully able to rebuild the application assets. Now I just need to figure out what the original problem was and how to fix it, but otherwise thanks for the help @krassowski!
